### PR TITLE
Enable group batch norm (--bnp) on ROCm (only bn_group = 1)

### PR DESCRIPTION
**Pull request by sarunyap**
Enable NHWC group batch norm on a single GPU on ROCm (bn_group = 1).  The
multi-GPU case (bn_group > 1) will be revisited in the future.

The following are the main changes:

1) Use MIOpen data structures/functions for HIP instead of CUDNN
2) For the warp-level primitive code for HIP, we ensure that the code operates
   on 64-thread wide warp instead of 32-thread wide
3) Disable all the bn_group > 1 paths for HIP
4) Make all data pointers contiguous to support non-contiguous input tensors
5) Use 64-bit ReLU bitmask for HIP
6) For fusion kernels, rectify the output data if it is less than or equal to
   zero to improve stability of the kernels

Notes:

1) Multi-stream is not tested.
2) We have not optimized for performance

### CODE REVIEW

File: apex/contrib/groupbn/batch_norm.py [hunk 1]
```diff
@@ -4,6 +4,16 @@
...
 
+def check_if_rocm_pytorch():
+    is_rocm_pytorch = False
+    if torch.__version__ >= '1.5':
+        from torch.utils.cpp_extension import ROCM_HOME
+        is_rocm_pytorch = True if ((torch.version.hip is not None) and (ROCM_HOME is not None)) else False
+
+    return is_rocm_pytorch
+
+IS_ROCM_PYTORCH = check_if_rocm_pytorch()
```
Comment (micmelesse) [inline apex/contrib/groupbn/batch_norm.py:15]: This a nitpick but could we unifiy or import this from the setup.py definition.

File: apex/contrib/csrc/groupbn/dnn.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#include <cudnn.h>
+#define DNN_STATUS_SUCCESS CUDNN_STATUS_SUCCESS
+#define DNN_DATA_HALF CUDNN_DATA_HALF
+#define DNN_TENSOR_FORMAT CUDNN_TENSOR_NHWC
+
+using dnnTensorFormat_t = cudnnTensorFormat_t;
+using dnnDataType_t = cudnnDataType_t;
+using dnnStatus_t = cudnnStatus_t;
+using dnnTensorDescriptor_t = cudnnTensorDescriptor_t;
+#endif
```
Comment (micmelesse) [inline apex/contrib/csrc/groupbn/dnn.h:24]: Ideally we would want hipify to handle all of these instead of introducing a new file. @jithunnair-amd What do you think?

File: apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h [hunk 3]
```diff
@@ -1289,23 +1533,27 @@ __global__ __launch_bounds__(THREADS_PER_CTA, DESIRED_OCCUPANCY)
...
+#endif
                     #pragma unroll
-                    for (int i = 0; i < ELEMENTS_PER_LDG; ++i) {
-                        bool rectified = x_math[i] < 0.0F;
-                        unsigned int local_relu_mask = __ballot_sync(0xFFFFFFFFU, rectified);
-                        if (lane_id == i) {
+                    for (int j = 0; j < ELEMENTS_PER_LDG; ++j) {
+                        bool rectified = __hle(__float2half(x_math[j]), zero_h);
+                        bitmask_t local_relu_mask = ballot(rectified);
+                        if (lane_id == j) {
```
Comment (micmelesse) [inline apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h:1546]: why not make this section of code an ifdef so that the path if preserved in cuda?

File: apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h [hunk 4]
```diff
@@ -1352,21 +1600,25 @@ __global__ __launch_bounds__(THREADS_PER_CTA, DESIRED_OCCUPANCY)
...
```
Comment (micmelesse) [inline apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h:1613]: ditto why not ifdef?

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h [hunk 5]
```diff
@@ -185,32 +199,44 @@ class NhwcBatchNormAddRelu {
...
     processCudnnStatus(status, "create tensor_descriptor");
   }
 
-  void destroyTensorDescriptor(cudnnTensorDescriptor_t descriptor) {
-    cudnnStatus_t status = CUDNN_STATUS_SUCCESS;
+  void destroyTensorDescriptor(dnnTensorDescriptor_t descriptor) {
+    dnnStatus_t status = DNN_STATUS_SUCCESS;
+#ifdef __HIP_PLATFORM_HCC__
+    status = miopenDestroyTensorDescriptor(descriptor);
+#else
```
Comment (micmelesse) [inline apex/contrib/csrc/groupbn/batch_norm_add_relu.h:229]: This should be handled by hipfiy. @jithunnair-amd

**Review by micmelesse — COMMENTED** [review]
This is great. Only issue that I see is that the hipfication is done using a lot of ifdefs even for stuff that should be handled by hipfiy. @jithunnair-amd Do you agree?

I think we can merge this if @sarunyap can add a test that proves correctness for the extension, the scripts you provided through email should be fine. We should probably add it to the ci as well.

I can come in later and clean it up to use hipfiy more and limit the ifdefs to chunks of code that could not hipified

File: apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h (hunk 3 above)
Comment (sarunyap) [reply apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h:1546]: Yes, that's a good idea.  I will fix this.

File: apex/contrib/groupbn/batch_norm.py (hunk 1 above)
Comment (sarunyap) [reply apex/contrib/groupbn/batch_norm.py:15]: I will look into this

**Comment by sarunyap** [comment]
> This is great. Only issue that I see is that the hipfication is done using a lot of ifdefs even for stuff that should be handled by hipfiy. @jithunnair-amd Do you agree?
> 
> I think we can merge this if @sarunyap can add a test that proves correctness for the extension, the scripts you provided through email should be fine. We should probably add it to the ci as well.
> 
> I can come in later and clean it up to use hipfiy more and limit the ifdefs to chunks of code that could not hipified

@micmelesse I have added a unit test for group BN.  Please let me know if you have more comments.  Thanks

File: apex/contrib/groupbn/batch_norm.py (hunk 1 above)
Comment (sarunyap) [reply apex/contrib/groupbn/batch_norm.py:15]: I have not refactored this code.  I think it will need a file restructuring in APEX.  So, I decided to leave this to the APEX team

File: apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h (hunk 3 above)
Comment (sarunyap) [reply apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h:1552]: I have fixed this

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h (hunk 5 above)
Comment (jithunnair-amd) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.h:229]: The MIOpen APIs are generally not signature-compatible with the CUDNN APIs, so we haven't added many of them to the hipify mappings. But I suppose we could add the ones that are compatible. I do agree that it shouldn't block this PR.

**Comment by micmelesse** [comment]
@sarunyap I tried the unit test and it looks good. 

There are some unit test failures. I am not quite sure if it is related to this PR or not. @hubertlu-tw Can you take a look at the failures and tell us if these failure are expected to fail.

File: apex/contrib/csrc/groupbn/cuda_utils.h [hunk 6]
```diff
@@ -1,4 +1,8 @@
+#ifdef __HIP_PLATFORM_HCC__
+#include <ATen/hip/HIPContext.h>
+#else
 #include <ATen/cuda/CUDAContext.h>
+#endif
 #ifndef CUDA_UTILS_H
```
Comment (jithunnair-amd) [inline apex/contrib/csrc/groupbn/cuda_utils.h:6]: This ifdef change is not needed since hipify will do this replacement in the generated hip_utils.h file

File: apex/contrib/csrc/groupbn/dnn.h (hunk 2 above)
Comment (jithunnair-amd) [reply apex/contrib/csrc/groupbn/dnn.h:24]: I think most of the above can be added to the hipify mappings file, except a couple like `CUDNN_TENSOR_NHWC`->`0` and `cudnnTensorFormat_t`->`int`. @sarunyap Could you please elaborate on these two mappings? Are they done because MIOpen doesn't have any equivalent types? Would these mappings be correct in the general case or only for the use-case here?

File: apex/contrib/csrc/groupbn/cuda_utils.h [hunk 7]
```diff
@@ -8,7 +12,11 @@ namespace cuda {
 namespace utils {
 
 static inline int MaxSharedMemoryPerMultiprocessor(int device_id) {
+#ifdef __HIP_PLATFORM_HCC__
+    return getDeviceProperties(device_id)->maxSharedMemoryPerMultiProcessor;
+#else
     return getDeviceProperties(device_id)->sharedMemPerMultiprocessor;
+#endif
 }
```
Comment (jithunnair-amd) [inline apex/contrib/csrc/groupbn/cuda_utils.h:20]: @micmelesse Do you see any risk in adding this mapping: `sharedMemPerMultiprocessor`->`maxSharedMemoryPerMultiProcessor` to the mappings file? It's not exactly a type or an api name, but it seems like it should be fine to add in the mappings file, even if it was just a variable name?

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.cu [hunk 8]
```diff
@@ -85,21 +85,23 @@ at::Tensor nhwc_bn_addrelu_fwd_train(
...
+  bn->setInputDescriptor(DNN_TENSOR_FORMAT, DNN_DATA_HALF, N, C, H, W, bn_group);
+  bn->setOutputDescriptor(DNN_TENSOR_FORMAT, DNN_DATA_HALF, N, C, H, W);
 
   bn->setConstants(momentum, epsilon);
 
   // set pointers within the wrapper
-  bn->setInputOutputPointers(x.DATA_PTR<at::Half>(),
+  bn->setInputOutputPointers(x.contiguous().DATA_PTR<at::Half>(),
                              nullptr,
                              y.DATA_PTR<at::Half>(),
```
Comment (jithunnair-amd) [inline apex/contrib/csrc/groupbn/batch_norm_add_relu.cu:96]: @sarunyap Could you please explain the reasoning behind using `.contiguous()` on all the tensors here? Does `y` not need to be contiguous?

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.cu [hunk 9]
```diff
@@ -167,21 +169,23 @@ at::Tensor nhwc_bn_addrelu_fwd_eval(
...
```
Comment (jithunnair-amd) [inline apex/contrib/csrc/groupbn/batch_norm_add_relu.cu:180]: @sarunyap Same question as previous about `y` not needing to be contiguous?

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h [hunk 10]
```diff
@@ -399,7 +528,11 @@ class NhwcBatchNormAddRelu {
   // Calculate the expected fwd kernel occupancy, as dictated by shared memory usage.
   static int smem_driven_fwd_occupancy(int device_id, const int max_cta_per_sm) {
     using namespace at::cuda::utils;
+#ifdef __HIP_PLATFORM_HCC__
+    int fwd_reduction_bytes = THREADS_PER_PIXEL*(THREADS_PER_CTA/64)*ELEMENTS_PER_LDG*sizeof(float);
+#else
     int fwd_reduction_bytes = THREADS_PER_PIXEL*(THREADS_PER_CTA/32)*ELEMENTS_PER_LDG*sizeof(float);
```
Comment (jithunnair-amd) [inline apex/contrib/csrc/groupbn/batch_norm_add_relu.h:534]: @sarunyap If this is due to the warp_size difference between Nvidia and AMD hardware, we should define a WARP_SIZE in a common header and use it everywhere we need warp-aware code, instead of hardcoding it this way.

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h [hunk 11]
```diff
@@ -408,7 +541,11 @@ class NhwcBatchNormAddRelu {
   // Calculate the expected bwd kernel occupancy, as dictated by shared memory usage.
   static int smem_driven_bwd_occupancy(int device_id, const int max_cta_per_sm) {
     using namespace at::cuda::utils;
+#ifdef __HIP_PLATFORM_HCC__
+    int bwd_reduction_bytes = THREADS_PER_PIXEL*(THREADS_PER_CTA/64)*ELEMENTS_PER_LDG*sizeof(float);
+#else
     int bwd_reduction_bytes = THREADS_PER_PIXEL*(THREADS_PER_CTA/32)*ELEMENTS_PER_LDG*sizeof(float);
+#endif
```
Comment (jithunnair-amd) [inline apex/contrib/csrc/groupbn/batch_norm_add_relu.h:548]: Ditto about WARP_SIZE

File: setup.py [hunk 12]
```diff
@@ -246,7 +246,8 @@ def check_cuda_torch_binary_vs_bare_metal(cuda_dir):
                                    'apex/contrib/csrc/groupbn/ipc.cu',
                                    'apex/contrib/csrc/groupbn/interface.cpp',
                                    'apex/contrib/csrc/groupbn/batch_norm_add_relu.cu'],
-                          include_dirs=[os.path.join(this_dir, 'csrc')],
+                          include_dirs=[os.path.join(this_dir, 'csrc'),
+                                        os.path.join(this_dir, 'apex/contrib/csrc/groupbn')],
                           extra_compile_args={'cxx': [] + version_dependent_macros,
```
Comment (jithunnair-amd) [inline setup.py:251]: @sarunyap This shouldn't be needed as hipify will automatically search in the current directory of the source file for the header?

**Review by jithunnair-amd — CHANGES_REQUESTED** [review]
@sarunyap @micmelesse This is a really big changeset, which makes it a bit challenging to review fully. However, a lot of the changes are due to ifdefing, as @micmelesse pointed out, which should be possible to handle using hipify. My preference would be to resolve any such ifdefs by leveraging hipify so that the PR changes are more manageable, in case we need to refer to it in the future when debugging an issue with this extension. However, in the interest of expediency, I'd be open to merging this PR into a separate branch of Apex instead of master branch. Would that be an acceptable short-term situation?

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.cu (hunk 8 above)
Comment (sarunyap) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.cu:96]: Tensor x is the "input" tensor that is given by a user which can be non-contiguous.  Tensor y is created inside the function which is always contiguous.  So, tensor y does not need to be made contiguous before being used.

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.cu (hunk 9 above)
Comment (sarunyap) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.cu:180]: Same reasoning:

Tensor x is the "input" tensor that is given by a user which can be non-contiguous.  Tensor y is created inside the function which is always contiguous.  So, tensor y does not need to be made contiguous before being used.

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h (hunk 10 above)
Comment (sarunyap) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.h:534]: Yes, this is due to the different warp sizes.  Does APEX or PyTorch already have warp size defined somewhere that I can use?

File: setup.py (hunk 12 above)
Comment (sarunyap) [reply setup.py:251]: I tried building APEX without this and I ran into a build problem.  So, I think it's needed.

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h (hunk 10 above)
Comment (jeffdaily) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.h:534]: PyTorch defines the macro C10_WARP_SIZE in c10/macros/Macros.h.  But hip_runtime.h defines `warpSize` as a constexpr.  It depends on if we want the code to be buildable for both AMD and NV.  `warpSize` is only defined for device code on NV and is not a constexpr so cannot be used for fixed-size arrays.

**Comment by sarunyap** [comment]
I'm not able to reply to https://github.com/ROCmSoftwarePlatform/apex/pull/51#discussion_r699870260 for some reason

@jithunnair-amd 

I think most of the above can be added to the hipify mappings file, except a couple like CUDNN_TENSOR_NHWC->0 and cudnnTensorFormat_t->int. @sarunyap Could you please elaborate on these two mappings? Are they done because MIOpen doesn't have any equivalent types? Would these mappings be correct in the general case or only for the use-case here?

So I'm replying here instead:

Our MIOpen has only one tensor format in the released version (as far as I know).  It does not support NHWC yet.  So, miopenTensorFormat_t and MIOPEN_TENSOR_NHWC do not exist yet.  That's why I mapped it to 0 and int.

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h (hunk 10 above)
Comment (sarunyap) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.h:534]: Thanks, Jeff.  So, I think it should be safe to use C10_WARP_SIZE?

Comment (jeffdaily) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.h:534]: Yes, you should be able to use C10_WARP_SIZE.  c10/macros/Macros.h is an installed header for pytorch.

File: setup.py (hunk 12 above)
Comment (jithunnair-amd) [reply setup.py:251]: @sarunyap I just merged #53 which works around an ordering issue during hipification of the various Apex extensions. Would you kindly rebase/merge your changes and try without the above entry to see if you still get a build issue?

**Comment by hubertlu-tw** [comment]
> @sarunyap I tried the unit test and it looks good.
> 
> There are some unit test failures. I am not quite sure if it is related to this PR or not. @hubertlu-tw Can you take a look at the failures and tell us if these failure are expected to fail.

I conducted A/B testing for this PR in the [rocm4.3_ubuntu18.04_py3.6_pytorch](https://hub.docker.com/layers/rocm/pytorch/rocm4.3_ubuntu18.04_py3.6_pytorch/images/sha256-1982afdafebfcc42eb6f1d6a7cde56240f738d93c481f9f7c9cd35d8e6233a94?context=explore) Docker container and compared with the current master branch.

=== **Test A** ===
Apex built from [ROCmSoftwarePlatform:master](https://github.com/ROCmSoftwarePlatform/apex) by the following command:
```
python3.6 setup.py install --cuda_ext --cpp_ext
```
The log file for building Apex can be found [apex_install_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094405/apex_install_A.txt).
The log files for [the unit tests ](https://confluence.amd.com/display/MLSE/Prototyping+APEX+on+ROCm#PrototypingAPEXonROCm-Unittests) can be found [apex_unittests_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094426/apex_unittests_A.txt) and [apex_distributed_unittests_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094430/apex_distributed_unittests_A.txt).


---
=== **Test B** ===
First, merged [sarunyap:group-bn-hip](https://github.com/sarunyap/apex/tree/group-bn-hip) and [ROCmSoftwarePlatform:master](https://github.com/ROCmSoftwarePlatform/apex). Then, built Apex by the following command:
```
python3.6 setup.py install --cuda_ext --cpp_ext --bnp
```
The log file for building Apex can be found [apex_install_bnp_merge_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094412/apex_install_bnp_merge_B.txt).
The log files for [the unit tests ](https://confluence.amd.com/display/MLSE/Prototyping+APEX+on+ROCm#PrototypingAPEXonROCm-Unittests) can be found [apex_unittests_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094440/apex_unittests_B.txt) and [apex_distributed_unittests_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094443/apex_distributed_unittests_B.txt).

In summary, the failures of the unit tests for this PR are expected to fail. Please correct me if I missed something for the A/B testing.

**Comment by sarunyap** [comment]
> > @sarunyap I tried the unit test and it looks good.
> > There are some unit test failures. I am not quite sure if it is related to this PR or not. @hubertlu-tw Can you take a look at the failures and tell us if these failure are expected to fail.
> 
> I conducted A/B testing for this PR in the [rocm4.3_ubuntu18.04_py3.6_pytorch](https://hub.docker.com/layers/rocm/pytorch/rocm4.3_ubuntu18.04_py3.6_pytorch/images/sha256-1982afdafebfcc42eb6f1d6a7cde56240f738d93c481f9f7c9cd35d8e6233a94?context=explore) Docker container and compared with the current master branch.
> 
> === **Test A** ===
> Apex built from [ROCmSoftwarePlatform:master](https://github.com/ROCmSoftwarePlatform/apex) by the following command:
> 
> ```
> python3.6 setup.py install --cuda_ext --cpp_ext
> ```
> 
> The log file for building Apex can be found [apex_install_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094405/apex_install_A.txt).
> The log files for [the unit tests ](https://confluence.amd.com/display/MLSE/Prototyping+APEX+on+ROCm#PrototypingAPEXonROCm-Unittests) can be found [apex_unittests_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094426/apex_unittests_A.txt) and [apex_distributed_unittests_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094430/apex_distributed_unittests_A.txt).
> 
> === **Test B** ===
> First, merged [sarunyap:group-bn-hip](https://github.com/sarunyap/apex/tree/group-bn-hip) and [ROCmSoftwarePlatform:master](https://github.com/ROCmSoftwarePlatform/apex). Then, built Apex by the following command:
> 
> ```
> python3.6 setup.py install --cuda_ext --cpp_ext --bnp
> ```
> 
> The log file for building Apex can be found [apex_install_bnp_merge_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094412/apex_install_bnp_merge_B.txt).
> The log files for [the unit tests ](https://confluence.amd.com/display/MLSE/Prototyping+APEX+on+ROCm#PrototypingAPEXonROCm-Unittests) can be found [apex_unittests_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094440/apex_unittests_B.txt) and [apex_distributed_unittests_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094443/apex_distributed_unittests_B.txt).
> 
> In summary, the failures of the unit tests for this PR are expected to fail. Please correct me if I missed something for the A/B testing.

If I understand correctly, the failed tests are not related to this PR, right?

**Comment by hubertlu-tw** [comment]
> > > @sarunyap I tried the unit test and it looks good.
> > > There are some unit test failures. I am not quite sure if it is related to this PR or not. @hubertlu-tw Can you take a look at the failures and tell us if these failure are expected to fail.
> > 
> > 
> > I conducted A/B testing for this PR in the [rocm4.3_ubuntu18.04_py3.6_pytorch](https://hub.docker.com/layers/rocm/pytorch/rocm4.3_ubuntu18.04_py3.6_pytorch/images/sha256-1982afdafebfcc42eb6f1d6a7cde56240f738d93c481f9f7c9cd35d8e6233a94?context=explore) Docker container and compared with the current master branch.
> > === **Test A** ===
> > Apex built from [ROCmSoftwarePlatform:master](https://github.com/ROCmSoftwarePlatform/apex) by the following command:
> > ```
> > python3.6 setup.py install --cuda_ext --cpp_ext
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The log file for building Apex can be found [apex_install_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094405/apex_install_A.txt).
> > The log files for [the unit tests ](https://confluence.amd.com/display/MLSE/Prototyping+APEX+on+ROCm#PrototypingAPEXonROCm-Unittests) can be found [apex_unittests_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094426/apex_unittests_A.txt) and [apex_distributed_unittests_A.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094430/apex_distributed_unittests_A.txt).
> > === **Test B** ===
> > First, merged [sarunyap:group-bn-hip](https://github.com/sarunyap/apex/tree/group-bn-hip) and [ROCmSoftwarePlatform:master](https://github.com/ROCmSoftwarePlatform/apex). Then, built Apex by the following command:
> > ```
> > python3.6 setup.py install --cuda_ext --cpp_ext --bnp
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The log file for building Apex can be found [apex_install_bnp_merge_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094412/apex_install_bnp_merge_B.txt).
> > The log files for [the unit tests ](https://confluence.amd.com/display/MLSE/Prototyping+APEX+on+ROCm#PrototypingAPEXonROCm-Unittests) can be found [apex_unittests_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094440/apex_unittests_B.txt) and [apex_distributed_unittests_B.txt](https://github.com/ROCmSoftwarePlatform/apex/files/7094443/apex_distributed_unittests_B.txt).
> > In summary, the failures of the unit tests for this PR are expected to fail. Please correct me if I missed something for the A/B testing.
> 
> If I understand correctly, the failed tests are not related to this PR, right?

Yes, the failed tests are not related to this PR.

**Comment by micmelesse** [comment]
@hubertlu-tw Thanks for confirming.

File: apex/contrib/csrc/groupbn/cuda_utils.h (hunk 7 above)
Comment (micmelesse) [reply apex/contrib/csrc/groupbn/cuda_utils.h:20]: It should be fine.

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h (hunk 10 above)
Comment (sarunyap) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.h]: @micmelesse @jithunnair-amd @jeffdaily I updated the code to use C10_WARP_SIZE

File: apex/contrib/csrc/groupbn/batch_norm_add_relu.h (hunk 11 above)
Comment (sarunyap) [reply apex/contrib/csrc/groupbn/batch_norm_add_relu.h]: @micmelesse @jithunnair-amd @jeffdaily I updated the code to use C10_WARP_SIZE here too

File: setup.py (hunk 12 above)
Comment (sarunyap) [reply setup.py:251]: @jithunnair-amd I tried this but it still gives me error

Comment (jithunnair-amd) [reply setup.py:251]: @sarunyap I dug into this, and you are right that the hipify logic doesn't look into the current directory of the `'apex/contrib/csrc/groupbn/batch_norm_add_relu.cu'` file for , so it doesn't find `batch_norm_add_relu.h` unless we explicitly add the `'apex/contrib/csrc/groupbn'` path to the `include_dirs`. This is actually a known bug in hipify where the `include_dirs` argument in setup.py call is erroneously passed in as the `includes` argument to hipify call. We'll need to fix this hipify bug in upstream pytorch. Until then, the change above should work.
